### PR TITLE
Add policy retention step for upgrade

### PR DESCRIPTION
**Pull request by hzur**
### Desired Outcome

We want to leave just 20 versions for each policy on upgrade.

### Implemented Changes

Added a migration step for removing old versions (all except the last 20)

### Connected Issue/Story

CyberArk internal issue link [ONYX-15331](https://ca-il-jira.il.cyber-ark.com:8443/secure/RapidBoard.jspa?rapidView=1475&view=detail&selectedIssue=ONYX-15331)

### Definition of Done
*At least 1 todo must be completed in the sections below for the PR to be
merged.*

#### Changelog

- [V ] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [ V] This PR includes new unit and integration tests to go with the code
  changes, or
- [ ] The changes in this PR do not require tests

#### Documentation

- [ ] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]()
- [ ] This PR does not require updating any documentation

#### Behavior

- [ ] This PR changes product behavior and has been reviewed by a PO, or
- [ ] These changes are part of a larger initiative that will be reviewed later, or
- [ ] No behavior was changed with this PR

#### Security

- [ ] Security architect has reviewed the changes in this PR,
- [ ] These changes are part of a larger initiative with a separate security review, or
- [ ] There are no security aspects to these changes

### CODE REVIEW

File: cucumber/api/features/policy_versions_retention.feature [hunk 1]
```diff
@@ -24,3 +24,13 @@ Feature: Limit the number of policy versions
     """
     Deleting policy version: {:version=>1, :resource_id=>"cucumber:policy:policy_test_version",
     """
+
+  Scenario: Migrate db remove extra policies
+    Given I save my place in the log file
+    And I successfully POST 21 times "/policies/cucumber/policy/policy_test_version" with body:
+    """
+      - !user bob
+    """    And the HTTP response status code is 201
```
Comment (orenbm) [inline cucumber/api/features/policy_versions_retention.feature:33]: move the And to a new line please

File: cucumber/api/features/policy_versions_retention.feature [hunk 2]
```diff
@@ -24,3 +24,13 @@ Feature: Limit the number of policy versions
...
+    Then I migrate the db
```
Comment (orenbm) [inline cucumber/api/features/policy_versions_retention.feature:34]: When?

File: cucumber/api/features/policy_versions_retention.feature [hunk 3]
```diff
@@ -24,3 +24,13 @@ Feature: Limit the number of policy versions
...
+    And I successfully GET "/resources/cucumber/policy/policy_test_version"
+    And there are 20 "policy_versions" in the response
```
Comment (orenbm) [inline cucumber/api/features/policy_versions_retention.feature:36]: Then?

File: cucumber/api/features/policy_versions_retention.feature [hunk 4]
```diff
@@ -24,3 +24,13 @@ Feature: Limit the number of policy versions
     """
     Deleting policy version: {:version=>1, :resource_id=>"cucumber:policy:policy_test_version",
     """
+
+  Scenario: Migrate db remove extra policies
```
Comment (orenbm) [inline cucumber/api/features/policy_versions_retention.feature:28]: do we want to also add this verification?
```
    And The following appears in the log after my savepoint:
    """
    Deleting policy version: {:version=>1, :resource_id=>"cucumber:policy:policy_test_version",
    """
```

Comment (hzur) [reply cucumber/api/features/policy_versions_retention.feature:28]: It shouldn't, the upgrade should not write to the log

File: cucumber/api/features/policy_versions_retention.feature [hunk 5]
```diff
@@ -24,3 +24,14 @@ Feature: Limit the number of policy versions
...
```
Comment (orenbm) [inline cucumber/api/features/policy_versions_retention.feature:28]:
```suggestion
  Scenario: DB migration removes policy versions that exceed to limit
```

File: cucumber/api/features/policy_versions_retention.feature [hunk 6]
```diff
@@ -24,3 +24,14 @@ Feature: Limit the number of policy versions
...
+    Given I save my place in the log file
+    And I successfully POST 21 times "/policies/cucumber/policy/policy_test_version" with body:
+    """
+      - !user bob
+    """    
+    And the HTTP response status code is 201
+    When I migrate the db
+    And I successfully GET "/resources/cucumber/policy/policy_test_version"
+    Then there are 20 "policy_versions" in the response
```
Comment (eladkug) [inline cucumber/api/features/policy_versions_retention.feature:37]: Please add some logs verification to the test to understand the job was done

Comment (hzur) [reply cucumber/api/features/policy_versions_retention.feature:37]: As far as I know-  upgrade (db migrate) does not written to the log in a testable matter, can you elaborate on how do you want to test it?

Comment (eladkug) [reply cucumber/api/features/policy_versions_retention.feature:37]: In our sanity we have something like this
`    And The following appears in the log after my savepoint:
    """
    Deleting policy version: {:version=>1, :resource_id=>"cucumber:policy:policy_test_version",
    """`
Something like this, as part of the upgrade we dont want to see this info on our logs ?

Comment (hzur) [reply cucumber/api/features/policy_versions_retention.feature:37]: no, because we're running just a SQL script for multiple deletions

Comment (eladkug) [reply cucumber/api/features/policy_versions_retention.feature:37]: @InbalZilberman WDYT? during the migrate do we want/need to have some logs?

**Review by orenbm — DISMISSED** [review]

**Review by orenbm — APPROVED** [review]

**Pull request merged**